src/usr/local/buildpack/utils/cache.sh

```diff
@@ -49,28 +58,107 @@ function get_from_url () {
       echo "${BUILDPACK_CACHE_DIR}/${filename}"
   else
     # cache disabled or not in cache
-    download_file "${url}" "${filename}"
+    download_file "${url}" "${filename}" "${expected_checksum}" "${checksum_algo}"
   fi
 }
 
 # Will download the file into the cache folder and returns the path
 # If the cache is not enabled it will download it to a temp folder
-# The second argument will be the filename if given
+# The second argument will be the filename if given (optional)
+# The third argument is the exepcted checksum of the file if given (optional)
+# The fourth argument is the checksum algorithm (optional)
 function download_file () {
   local url=${1}
   check url true
 
   local name
   name=${2:-$(basename "${url}")}
 
-  local temp_folder=${BUILDPACK_CACHE_DIR:-${TEMP_DIR}}
-  if ! curl --retry 3 --create-dirs -sSfLo "${temp_folder}/${name}" "${url}" ; then
-    echo "Download failed: ${url}" >&2
-    exit 1
-  fi;
+  local expected_checksum
+  expected_checksum=${3}
+
+  local checksum_algo
+  checksum_algo=${4}
+
+  local retry=3
+  while [ "${retry}" -gt 0 ]; do
+    retry=$((retry-1))
+    local temp_folder=${BUILDPACK_CACHE_DIR:-${TEMP_DIR}}
```

Comment:
shouldn't this be outside the loop because it's static?

Reply:
It should, will move code around